test/integration/stats_integration_test.cc

```diff
@@ -369,6 +373,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeHostSizeWithStats) {
   // 2019/11/01  8859     1299         1315   build: switch to libc++ by default
   // 2019/11/12  8998     1299         1350   test: adjust memory limit for macOS
   // 2019/11/15  9040     1283         1350   build: update protobuf to 3.10.1
+  // 2020/01/13  9663     1619         1655   api: deprecate hosts in Cluster.
```

Comment:
Yikes. cc @jmarantz. Probably worth a follow up issue here to look into this.

Reply:
In thinking about this more I'm a little confused why this change effects the per-host size? Is any of this data stored on a per-host basis?

Reply:
I have done some analysis on this (using a pre #9633 branch to avoid any possible interference from that stuff). I have a simple benchmark script at https://github.com/htuch/envoy/blob/cla-overhead/test/common/upstream/host_overhead_test.cc that just measures the protobuf overhead difference between the two host data structures.

The results look like:
```
hosts structure overhead per host: 106.568
load_assignment structure overhead per host: 218.572
hosts overhead per host: 138.568
load_assignment overhead per host: 250.572
```

The load assignment comes at the cost of an extra 112 bytes per host due to the additional proto structure (when configured similarly to `stats_integration_test`). We have two obvious copies of each host being measured by the benchmark, one for the `bootstrap`, which is still around at the end of the test and another in the cluster's config dump copy, since these are inlined in `Cluster`, so that takes us to 224 bytes.  The above regression is 336 bytes, which strongly suggest via interpolation that we have a third copy somewhere.

So, I think the overhead here is purely explicable by the more deeply nested structure. Not sure where we want to go in this PR or beyond. Ideas:
* We could exclude the inlined hosts from the cluster proto copy kept for config dump, since it's somewhat redundant with /clusters. However, we want to add EDS to config dump in the future, so maybe not.
* We probably should try and avoid including the input config `boostrap` in the measured part of `stats_integration_test`. If we do that, we 're changing our methodology, so would need to note that in the measurement history.

Reply:
In the actual product, do we keep the input protobufs around after the xDS events are processed? If not, then we should also make sure we drop them in the memory integration test. I can follow up on that.

Reply:
We keep copies for config dump purposes. The original bootstrap is also kept in `Server::InstanceImpl`. That probably isn't needed post-init, so we could in theory save memory there by removing. I'm not sure how practically useful it is, since I imagine most of the scale-up endpoint scenarios are dynamic, so you would only be making improvements for synthetic situations.